API: Make item `Visibility` visible in the API

xFrednet opened the pull request:
Designing a sensible `Visibility` representation is surprisingly hard. I had to update the function names several times, and still feel like they can be improved. There is also still one util function missing, which requires the implementation of https://github.com/rust-marker/marker/issues/242 first. I've left a corresponding comment in the PR.

I think this implementation will cover 99% of the current use cases. This PR also updates the `not_using_has_span_trait` lint and enables it by default inside the `marker_api` crate

---

Closes: rust-marker/marker#26

Veetaha commented on marker_api/src/ast/item.rs:
```diff
@@ -238,42 +240,182 @@ use impl_item_data;
-            vis: Visibility::new(id),
-            ident,
-        }
+    // Here we can't `derive(TypedBuilder)`, as the created builder is private,
+    // due to the declaration of the item not being pub.
+    pub fn new(id: ItemId, span: SpanId, vis: Visibility<'ast>, ident: Ident<'ast>) -> Self {
```
This surprised me, so I checked and the following works. You just need to make sure that `visibility::make(pub)` preceds the `TypedBuilder` derive due to the order of macro expansion.
```rust
#[repr(C)]
#[derive(Debug)]
#[cfg_attr(feature = "driver-api", visibility::make(pub))]
#[cfg_attr(feature = "driver-api", derive(typed_builder::TypedBuilder))]
struct CommonItemData<'ast> {
    id: ItemId,
    span: SpanId,
    vis: Visibility<'ast>,
    ident: Ident<'ast>,
}
```

Here, for example, it already works like that.
https://github.com/rust-marker/marker/blob/73835de0a82ebd33b959df200f66e5a3d7bf007a/marker_api/src/ast/stmt.rs#L59-L73

Veetaha reviewed (commented):
I'll take a closer look on weekends, didn't review everything yet.

Veetaha commented on marker_api/src/ast/item.rs:
```diff
@@ -238,42 +240,182 @@ use impl_item_data;
+    ///
+    /// This function checks if the visibility is restricted and the defined path
+    /// belongs to the root module of the crate. Meaning, that this can also be `true`,
+    /// if the visibility uses `pub(super)` to travel up to the crate root.
```
So this will return true if the path refers to the root of the crate. That looks more like the semantic property of this type, not the ast. Maybe it makes sense to make the ast/sem distinction more exlicit? Is it even possible to check if the visibility is `pub(crate)` syntactically (and not `pub(super)` or `pub(in ...)` that lead to the root of the crate).

xFrednet replied:
Rustc sadly erases the information how the path has been specified. That's why this type provides semantic information when it comes to this. I could rearrange it, so that we have a semantic visibility, and the AST visibility, basically just wraps that one and optionally provides a `Span`. What do you think about this structure?

Veetaha replied:
I see, that's a pity. Anyway, since we don't have a different from the original AST representation of the visibility, I think it does make sense

Veetaha commented on marker_api/src/ast/item.rs:
````diff
@@ -238,42 +240,182 @@ use impl_item_data;
+    /// belongs to the root module of the crate. Meaning, that this can also be `true`,
+    /// if the visibility uses `pub(super)` to travel up to the crate root.
+    ///
+    /// ```rs
````
`rs` disables the code highlighting of in the doc comment in rust-analyzer, and I think `cargo test --doc` will also not try to compile this code as a doc test. Just use bare triple-tick

```suggestion
    /// ```
```

Veetaha commented on marker_api/src/ast/item.rs:
```diff
@@ -238,42 +240,182 @@ use impl_item_data;
+    ///
+    ///     // Returns `true` as the visibility is restricted to `super` which is the
+    ///     // root of the crate.
+    ///     pub(crate) boo() {}
```
```suggestion
    ///     pub(super) boo() {}
```

Veetaha commented on marker_api/src/ast/item.rs:
```diff
@@ -238,42 +240,182 @@ use impl_item_data;
+    ///     pub(crate) boo() {}
+    /// }
+    ///
+    /// // Returns `false` since the visibility is not restricted
+    /// fn example_in_root() {}
```
This makes it again special case. I don't like how syntactic and semantic meaning of `pub(crate)` are intermixed here.
What I would expect is smth like this:
- `ast::Visibility::is_pub_crate()`: checks if the visibility is spelled exactly as `pub(crate)` returning false for any other syntax

- `ast::Visibility::is_pub_in_path()`: checks if the visibility is spelled exactly as `pub(in ...)` returning false for any other syntax

- `sem::Visibility::is_pub_crate()`: checks if the visibility makes the symbol accessible in the crate root (`pub(crate)`, `pub(super)`, `pub(in ...)`, default visibility in `lib.rs` all would return `true`)

- `sem::Visibility::is_pub_in_path(path)`: checks if the visibility makes the symbol accessible under the given path (any syntax e.g. `pub(crate)`, `pub(in ...)` may return `true` only if the resulting resolved visibility allows the symbol to be accessed in `path` module).

xFrednet replied:
See: https://github.com/rust-marker/marker/pull/294#discussion_r1375321200

Veetaha commented on marker_api/src/ast/item.rs:
```diff
@@ -238,42 +240,182 @@ use impl_item_data;
+        matches!(self.kind, VisibilityKind::Crate(_))
+    }
+
+    /// Returns `true` if a visibility has been defined.
+    pub fn is_defined(&self) -> bool {
+        !matches!(self.kind, VisibilityKind::Default(_) | VisibilityKind::DefaultPub)
+    }
```
The term "defined" implies there is a case where the visibility is "undefined", which reads to me as if "we just don't know what the visibility is if it's not defined". I think `is_non_default` or is `is_explicit` would read better.

Veetaha commented on marker_api/src/ast/item.rs:
```diff
@@ -238,42 +240,182 @@ use impl_item_data;
+    ///
+    /// See [`Visibility::module_id`] to get the `ItemId`, even if the item or
+    /// field uses the default visibility.
+    pub fn pub_with_path_module_id(&self) -> Option<ItemId> {
```
This again looks weird due to the mixture of syntactic and semantic properties of the visibility.

I think if we split the representation into `ast` and `sem` there would be only a method on `sem` type:

- `sem::Visibility::visibility_root() -> Option<ItemId>`: returns the `ItemId` of the root module this visibility level allows access in; returns `None` if the visibility is `pub`

And there wouldn't be such a method for `ast::Visibility`. I think it would be more straightforward to just expose the syntactic `VisibilityKind` to them, and the users would then handle all possible syntactic flavors of the visibility they want to check for.
But the `VisibilityKind` in the `ast` module would look like this:

```rust
enum VisibilityKind {
  Pub,
  PubIn(Path),
  PubCrate,
  
  /// There is no distinction between `Default/DefautPub` because in the synatctic representation
  /// we only care if the visibility was explicitly specified or not
  Default, 
}
```

Maybe I'm dreaming here, and rustc doesn't provide the original syntax form of the visibility, in which case it would be a bummer.

**UPD**: I noticed the `module_id` method only after writing this comment.

Veetaha commented on marker_rustc_driver/src/conversion/marker/ast/item.rs:
```diff
@@ -159,6 +165,26 @@ impl<'ast, 'tcx> MarkerConverterInner<'ast, 'tcx> {
     }
 
+    fn to_visibility(&self, owner_id: hir::def_id::LocalDefId, vis_span: rustc_span::Span) -> Visibility<'ast> {
+        let span = if vis_span.is_empty() {
+            None
+        } else {
+            Some(self.to_span_id(vis_span))
+        };
```
```suggestion
        let span = (!vis_span.is_empty()).then(|| self.to_span_id(vis_span));
```

Veetaha commented on marker_rustc_driver/src/conversion/marker/ast/item.rs:
```diff
@@ -249,7 +275,7 @@ impl<'ast, 'tcx> MarkerConverterInner<'ast, 'tcx> {
             // field after the next sync. See #55
```
This fixme comment may be removed already, right?

Veetaha commented on marker_uilints/src/lib.rs:
```diff
@@ -144,6 +149,26 @@ impl LintPass for TestLintPass {
                         diag.note(format!("Body: {:#?}", cx.ast().body(func.body_id().unwrap())));
                     });
             }
+            if matches!(
+                item.ident().map(marker_api::span::Ident::name),
+                Some(name) if name.starts_with("test_vis")
+            ) {
```
```suggestion
            if item.ident().is_some_and(|ident| ident.name().starts_with("test_vis")) {
```

xFrednet commented on marker_api/src/ast/item.rs:
```diff
@@ -238,42 +240,182 @@ use impl_item_data;
-/// This struct represents the visibility of an item.
+/// The declared visibility of an item or field.
+///
+/// Note that this is only the syntactic visibility. The item or field might be
+/// reexported with a higher visibility, or have a high default visibility.
```
That's really interesting, I never knew. Thank you!

xFrednet commented:
This is still WIP, so far I've only applied the easy changes and skimmed over your suggestions. I'll answer properly soon. Thank you for the review!

xFrednet commented:
Okay, I've rebased. Only the last two commits are new. I've now split the visibility into a semantic and ast representation, as discussed previously. Sorry that it took so long, to get this fixed :sweat_smile: Hope you have a beautiful weekend :D

Veetaha commented on marker_api/src/sem/item.rs:
```diff
@@ -0,0 +1,160 @@
+    /// This function checks if the visibility is restricted and the defined path
+    /// belongs to the root module of the crate. Meaning, that this can also be `true`,
+    /// if the visibility uses `pub(super)` to travel up to the crate root.
+    // Ignore, since the `in crate::example_1` path doesn't work for doc tests
```
I wonder how this code is layed out during compilation. That's super-unintuitive.
I've checked some metadata:
```
[marker_api/src/sem/item.rs:27] file!() = "marker_api/src/sem/item.rs"
[marker_api/src/sem/item.rs:28] module_path!() = "rust_out"
[marker_api/src/sem/item.rs:8] module_path!() = "rust_out::example_1"
thread 'main' panicked at marker_api/src/sem/item.rs:30:1:
rust_out::main::_doctest_main_marker_api_src_sem_item_rs_64_0::example_1::Baz
```
The panic mentions the `type_name` of a struct defined inside of an `example_1`. I guess rustdoc generates just a single crate for many dotctests if the code example doesn't include an explicit `main` function.

However, if you add an explicit `main()` function the path to `Baz` becomes this:
```
thread 'main' panicked at marker_api/src/sem/item.rs:4:5:
rust_out::example_1::Baz
```

So to remove the `ignore` in this doctest you can add an explicit `# fn main() {}` somewhere here. This will make `crate::example_1` work.

xFrednet replied:
That works perfectly, thank you! :heart:

Veetaha commented on marker_api/src/sem/item.rs:
```diff
@@ -0,0 +1,160 @@
+    ///     pub(super) fn boo() {}
+    /// }
+    ///
+    /// // Returns `false` since the visibility is not restricted
```
Technically, this item is also `pub(crate)`. That's a bit unintuitive. Could we identify that the default visibility at the root of the crate as implicitly `pub(crate)` as well? I guess it would require just another `VisibilityKind`

xFrednet replied:
Yes we could, and I would actually guess that this is currently wrongly classified as `pub(crate)`.

Veetaha commented on marker_api/src/sem/item.rs:
````diff
@@ -0,0 +1,160 @@
+    /// // Returns `false` since the visibility is not restricted
+    /// fn example_in_root() {}
+    /// ```
+    pub fn is_pub_crate(&self) -> bool {
````
I think people will be confused if we name this method after the special `pub(crate)` syntax when this method covers more than just that.

```suggestion
    pub fn is_crate_scoped(&self) -> bool {
```

Given the amount of info consider this number of methods instead:
- `is_crate_scoped() -> bool` (true if `scope` resolves to the current crate no matter what syntax was used)
- `is_pub() -> bool { self.scope().is_none() }`
- `is_default() -> bool`
- `scope() -> Option<ItemId>`
If the users wants to check if the visibility was specified (non-default), they would use `!is_default()`. If they want to exclude `pub` they will use `!is_default && !is_pub()`.

In other words I'd like to make it explicit that the semantic visibility no longer represents the original syntax that was used.

You can also notice that in the list of methods I suggested there isn't `pub_with_path_module_id`. I think it's not needed because it's expressible with just `vis.scope().filter(|_| !vis.is_default())` or similar.

xFrednet replied:
I really like the suggested names. I'll make the changes. We should be able to reparse at least part of the text snippet. I haven't worked with Rustc's lexer and parser yet, which is partially why I'm avoiding such solutions :sweat_smile:

Veetaha replied:
Re-parsing the text snippet is more of a hack. I wonder how clippy fights with this? Maybe other kinds of lint passes such as `EarlyLintPass` have the necessary ast info?

xFrednet replied:
Clippy usually just checks if things are public, and that's it. In Clippy we don't really have lints that span over more than a module and those that do don't require the visibility AFAIK.

For the `EarlyLintPass`, I've just checked and that one still has the path [[1](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/ast/enum.VisibilityKind.html)]. The big problem with this approach is that nodes from the AST, are hard to connect to the nodes from the HIR, as the AST doesn't have a unique ID for everything. `HirId`s are also structured differently.

The current rustc interface has covered all of Clippy's use cases so far. So I hope that this is enough for Marker.

Veetaha reviewed (commented):
Consider making design a bit more minimalistic but obvious as suggested in my comment.

xFrednet commented:
Okay, this should hopefully be it. Thank you for the excellent review. I think this very is way better, than my initial proposal :)

Veetaha commented on marker_api/src/sem/item.rs:
```diff
@@ -80,57 +75,52 @@ impl<'ast> Visibility<'ast> {
+            | VisibilityKind::Crate(id)
+            | VisibilityKind::DefaultCrate(id)
+            | VisibilityKind::Default(id) => Some(id),
             _ => None,
```
`[nit]`
I'd prefer an exhaustive match to make sure all cases are handled if `VisibilityKind` enum is modified

Veetaha reviewed (approved):
LGTM. Design is hard, but things should go on 👍
